File: packages/mobile-android/app/src/main/java/com/vangarments/navigation/NavigationManager.kt
package com.vangarments.navigation

import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import android.util.Log
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.State
import androidx.navigation.NavController
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.lang.ref.WeakReference

// Navigation Route data class
@Serializable
data class NavigationRoute(
    val name: String,
    val displayName: String,
    val icon: String,
    val requiresAuth: Boolean = false,
    val adminOnly: Boolean = false
)

// Navigation State data class
@Serializable
data class NavigationState(
    val currentRoute: String? = null,
    val previousRoute: String? = null,
    val navigationHistory: List<String> = emptyList(),
    val isNavigating: Boolean = false
)

class NavigationManager private constructor(private val context: Context) {
    
    companion object {
        private const val TAG = "NavigationManager"
        private const val PREFS_NAME = "navigation_prefs"
        private const val STATE_KEY = "navigation_state"
        private const val MAX_HISTORY_COUNT = 10
        
        @Volatile
        private var INSTANCE: NavigationManager? = null
        
        fun getInstance(context: Context): NavigationManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: NavigationManager(context.applicationContext).also { INSTANCE = it }
            }
        }
    }
    
    // App routes definition
    val appRoutes = listOf(
        NavigationRoute("wardrobe", "Guarda-roupa", "shirt", requiresAuth = true),
        NavigationRoute("discover", "Descobrir", "sparkles", requiresAuth = true),
        NavigationRoute("camera", "Adicionar", "camera", requiresAuth = true),
        NavigationRoute("marketplace", "Marketplace", "storefront", requiresAuth = true),
        NavigationRoute("profile", "Perfil", "person", requiresAuth = true),
        NavigationRoute("auth", "Autenticação", "login", requiresAuth = false)
    )
    
    private val _navigationState = MutableStateFlow(NavigationState())
    val navigationState: StateFlow<NavigationState> = _navigationState.asStateFlow()
    
    private val _selectedTab = mutableStateOf(0)
    val selectedTab: State<Int> = _selectedTab
    
    private var navControllerRef: WeakReference<NavController>? = null
    private val sharedPrefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    
    init {
        loadNavigationState()
        Log.d(TAG, "NavigationManager initialized")
    }
    
    // Set NavController reference
    fun setNavController(navController: NavController) {
        navControllerRef = WeakReference(navController)
        Log.d(TAG, "NavController set")
    }
    
    // Navigate to a route
    fun navigate(routeName: String, animated: Boolean = true): Boolean {
        return try {
            Log.d(TAG, "Navigating to: $routeName")
            
            if (!canNavigateToRoute(routeName)) {
                Log.w(TAG, "Cannot navigate to $routeName - access denied")
                return false
            }
            
            updateCurrentRoute(routeName)
            
            // Handle tab navigation
            getTabIndex(routeName)?.let { tabIndex ->
                _selectedTab.value = tabIndex
            }
            
            // Navigate using NavController if available
            navControllerRef?.get()?.let { navController ->
                try {
                    navController.navigate(routeName)
                } catch (e: Exception) {
                    Log.e(TAG, "NavController navigation failed", e)
                }
            }
            
            saveNavigationState()
            Log.d(TAG, "Navigation to $routeName successful")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Navigation failed", e)
            false
        }
    }
    
    // Go back
    fun goBack(): Boolean {
        return try {
            Log.d(TAG, "Going back")
            
            val currentState = _navigationState.value
            val targetRoute = currentState.previousRoute 
                ?: currentState.navigationHistory.getOrNull(currentState.navigationHistory.size - 2)
                ?: "wardrobe" // Default fallback
            
            navigate(targetRoute, animated = true)
        } catch (e: Exception) {
            Log.e(TAG, "Go back failed", e)
            false
        }
    }
    
    // Reset navigation
    fun reset(routeName: String): Boolean {
        return try {
            Log.d(TAG, "Resetting to: $routeName")
            
            _navigationState.value = NavigationState(
                currentRoute = routeName,
                previousRoute = null,
                navigationHistory = listOf(routeName),
                isNavigating = false
            )
            
            getTabIndex(routeName)?.let { tabIndex ->
                _selectedTab.value = tabIndex
            }
            
            saveNavigationState()
            Log.d(TAG, "Reset to $routeName successful")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Reset failed", e)
            false
        }
    }
    
    // Route management
    fun getRoute(name: String): NavigationRoute? {
        return appRoutes.find { it.name == name }
    }
    
    fun getAuthenticatedRoutes(): List<NavigationRoute> {
        return appRoutes.filter { it.requiresAuth }
    }
    
    fun getPublicRoutes(): List<NavigationRoute> {
        return appRoutes.filter { !it.requiresAuth }
    }
    
    fun canNavigateToRoute(
        routeName: String, 
        isAuthenticated: Boolean = true, 
        isAdmin: Boolean = false
    ): Boolean {
        val route = getRoute(routeName) ?: return true // Allow unknown routes
        
        if (route.adminOnly && !isAdmin) return false
        if (route.requiresAuth && !isAuthenticated) return false
        
        return true
    }
    
    // Private methods
    private fun updateCurrentRoute(routeName: String) {
        val currentState = _navigationState.value
        val newHistory = (currentState.navigationHistory + routeName)
            .takeLast(MAX_HISTORY_COUNT)
        
        _navigationState.value = currentState.copy(
            previousRoute = currentState.currentRoute,
            currentRoute = routeName,
            navigationHistory = newHistory
        )
        
        Log.d(TAG, "Route updated from ${currentState.currentRoute} to $routeName")
    }
    
    private fun getTabIndex(routeName: String): Int? {
        return when (routeName) {
            "wardrobe" -> 0
            "discover" -> 1
            "camera" -> 2
            "marketplace" -> 3
            "profile" -> 4
            else -> null
        }
    }
    
    // Persistence
    private fun saveNavigationState() {
        try {
            val json = Json.encodeToString(_navigationState.value)
            sharedPrefs.edit().putString(STATE_KEY, json).apply()
            Log.d(TAG, "Navigation state saved")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save navigation state", e)
        }
    }
    
    private fun loadNavigationState() {
        try {
            val json = sharedPrefs.getString(STATE_KEY, null)
            if (json != null) {
                val state = Json.decodeFromString<NavigationState>(json)
                _navigationState.value = state
                
                // Update selected tab
                state.currentRoute?.let { currentRoute ->
                    getTabIndex(currentRoute)?.let { tabIndex ->
                        _selectedTab.value = tabIndex
                    }
                }
                
                Log.d(TAG, "Navigation state loaded - current route: ${state.currentRoute}")
            } else {
                Log.d(TAG, "No saved navigation state found")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load navigation state", e)
        }
    }
    
    // Deep link handling
    fun handleDeepLink(uri: Uri): Boolean {
        return try {
            Log.d(TAG, "Handling deep link: $uri")
            
            val pathSegments = uri.pathSegments
            if (pathSegments.isNotEmpty()) {
                val routeName = pathSegments[0]
                navigate(routeName, animated = true)
            } else {
                Log.w(TAG, "Invalid deep link - no path segments")
                false
            }
        } catch (e: Exception) {
            Log.e(TAG, "Deep link handling failed", e)
            false
        }
    }
    
    // Debug
    fun debugNavigation() {
        val state = _navigationState.value
        Log.d(TAG, "=== Navigation Debug Info ===")
        Log.d(TAG, "Current Route: ${state.currentRoute}")
        Log.d(TAG, "Previous Route: ${state.previousRoute}")
        Log.d(TAG, "History: ${state.navigationHistory}")
        Log.d(TAG, "Selected Tab: ${_selectedTab.value}")
        Log.d(TAG, "Available Routes: ${appRoutes.size}")
        Log.d(TAG, "NavController Available: ${navControllerRef?.get() != null}")
        Log.d(TAG, "=============================")
    }
}

File: packages/mobile-ios/Navigation/NavigationManager.swift
import SwiftUI
import Combine

// MARK: - Navigation Route
struct NavigationRoute: Identifiable, Hashable {
    let id = UUID()
    let name: String
    let displayName: String
    let icon: String
    let requiresAuth: Bool
    let adminOnly: Bool
    
    init(name: String, displayName: String, icon: String, requiresAuth: Bool = false, adminOnly: Bool = false) {
        self.name = name
        self.displayName = displayName
        self.icon = icon
        self.requiresAuth = requiresAuth
        self.adminOnly = adminOnly
    }
}

// MARK: - Navigation State
struct NavigationState {
    var currentRoute: String?
    var previousRoute: String?
    var navigationHistory: [String]
    var isNavigating: Bool
    
    init() {
        self.currentRoute = nil
        self.previousRoute = nil
        self.navigationHistory = []
        self.isNavigating = false
    }
}

// MARK: - Navigation Manager
class NavigationManager: ObservableObject {
    static let shared = NavigationManager()
    
    @Published var navigationState = NavigationState()
    @Published var selectedTab = 0
    
    private let maxHistoryCount = 10
    
    // Define app routes
    let appRoutes: [NavigationRoute] = [
        NavigationRoute(name: "wardrobe", displayName: "Guarda-roupa", icon: "tshirt", requiresAuth: true),
        NavigationRoute(name: "discover", displayName: "Descobrir", icon: "sparkles", requiresAuth: true),
        NavigationRoute(name: "camera", displayName: "Adicionar", icon: "camera", requiresAuth: true),
        NavigationRoute(name: "community", displayName: "Comunidade", icon: "person.2", requiresAuth: true),
        NavigationRoute(name: "profile", displayName: "Perfil", icon: "person.circle", requiresAuth: true),
        NavigationRoute(name: "auth", displayName: "Autenticação", icon: "person.badge.key", requiresAuth: false)
    ]
    
    private init() {
        loadNavigationState()
    }
    
    // MARK: - Navigation Methods
    
    func navigate(to routeName: String, animated: Bool = true) {
        print("🔧 iOS Navigation: Navigating to \(routeName)")
        
        guard canNavigateToRoute(routeName) else {
            print("❌ iOS Navigation: Cannot navigate to \(routeName) - access denied")
            return
        }
        
        updateCurrentRoute(routeName)
        
        // Handle tab navigation
        if let tabIndex = getTabIndex(for: routeName) {
            withAnimation(animated ? .easeInOut(duration: 0.3) : nil) {
                selectedTab = tabIndex
            }
        }
        
        saveNavigationState()
        print("✅ iOS Navigation: Navigation to \(routeName) successful")
    }
    
    func goBack() {
        print("🔧 iOS Navigation: Going back")
        
        if let previousRoute = navigationState.previousRoute {
            navigate(to: previousRoute, animated: true)
        } else if !navigationState.navigationHistory.isEmpty {
            let lastRoute = navigationState.navigationHistory[navigationState.navigationHistory.count - 2]
            navigate(to: lastRoute, animated: true)
        } else {
            // Default to wardrobe if no history
            navigate(to: "wardrobe", animated: true)
        }
    }
    
    func reset(to routeName: String) {
        print("🔧 iOS Navigation: Resetting to \(routeName)")
        
        navigationState.navigationHistory.removeAll()
        navigationState.previousRoute = nil
        navigate(to: routeName, animated: false)
    }
    
    // MARK: - Route Management
    
    func getRoute(by name: String) -> NavigationRoute? {
        return appRoutes.first { $0.name == name }
    }
    
    func getAuthenticatedRoutes() -> [NavigationRoute] {
        return appRoutes.filter { $0.requiresAuth }
    }
    
    func getPublicRoutes() -> [NavigationRoute] {
        return appRoutes.filter { !$0.requiresAuth }
    }
    
    func canNavigateToRoute(_ routeName: String, isAuthenticated: Bool = true, isAdmin: Bool = false) -> Bool {
        guard let route = getRoute(by: routeName) else {
            return true // Allow unknown routes
        }
        
        if route.adminOnly && !isAdmin {
            return false
        }
        
        if route.requiresAuth && !isAuthenticated {
            return false
        }
        
        return true
    }
    
    // MARK: - Private Methods
    
    private func updateCurrentRoute(_ routeName: String) {
        let previousRoute = navigationState.currentRoute
        
        navigationState.previousRoute = previousRoute
        navigationState.currentRoute = routeName
        
        // Add to history
        navigationState.navigationHistory.append(routeName)
        
        // Limit history size
        if navigationState.navigationHistory.count > maxHistoryCount {
            navigationState.navigationHistory.removeFirst()
        }
        
        print("🔧 iOS Navigation: Route updated from \(previousRoute ?? "nil") to \(routeName)")
    }
    
    private func getTabIndex(for routeName: String) -> Int? {
        switch routeName {
        case "wardrobe":
            return 0
        case "discover":
            return 1
        case "camera":
            return 2
        case "community":
            return 3
        case "profile":
            return 4
        default:
            return nil
        }
    }
    
    // MARK: - Persistence
    
    private func saveNavigationState() {
        do {
            let data = try JSONEncoder().encode(NavigationStateData(from: navigationState))
            UserDefaults.standard.set(data, forKey: "navigationState")
            print("✅ iOS Navigation: State saved")
        } catch {
            print("❌ iOS Navigation: Failed to save state - \(error)")
        }
    }
    
    private func loadNavigationState() {
        guard let data = UserDefaults.standard.data(forKey: "navigationState") else {
            print("🔧 iOS Navigation: No saved state found")
            return
        }
        
        do {
            let stateData = try JSONDecoder().decode(NavigationStateData.self, from: data)
            navigationState = stateData.toNavigationState()
            
            // Update selected tab if current route is available
            if let currentRoute = navigationState.currentRoute,
               let tabIndex = getTabIndex(for: currentRoute) {
                selectedTab = tabIndex
            }
            
            print("✅ iOS Navigation: State loaded - current route: \(navigationState.currentRoute ?? "nil")")
        } catch {
            print("❌ iOS Navigation: Failed to load state - \(error)")
        }
    }
    
    // MARK: - Deep Link Handling
    
    func handleDeepLink(_ url: URL) -> Bool {
        print("🔧 iOS Navigation: Handling deep link - \(url)")
        
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        
        guard !pathComponents.isEmpty else {
            print("❌ iOS Navigation: Invalid deep link - no path components")
            return false
        }
        
        let routeName = pathComponents[0]
        navigate(to: routeName, animated: true)
        
        return true
    }
    
    // MARK: - Debug
    
    func debugNavigation() {
        print("🔧 iOS Navigation Debug Info:")
        print("  Current Route: \(navigationState.currentRoute ?? "nil")")
        print("  Previous Route: \(navigationState.previousRoute ?? "nil")")
        print("  History: \(navigationState.navigationHistory)")
        print("  Selected Tab: \(selectedTab)")
        print("  Available Routes: \(appRoutes.count)")
    }
}

// MARK: - Codable Support
private struct NavigationStateData: Codable {
    let currentRoute: String?
    let previousRoute: String?
    let navigationHistory: [String]
    let isNavigating: Bool
    
    init(from state: NavigationState) {
        self.currentRoute = state.currentRoute
        self.previousRoute = state.previousRoute
        self.navigationHistory = state.navigationHistory
        self.isNavigating = state.isNavigating
    }
    
    func toNavigationState() -> NavigationState {
        var state = NavigationState()
        state.currentRoute = currentRoute
        state.previousRoute = previousRoute
        state.navigationHistory = navigationHistory
        state.isNavigating = isNavigating
        return state
    }
}

// MARK: - SwiftUI Environment
struct NavigationManagerKey: EnvironmentKey {
    static let defaultValue = NavigationManager.shared
}

extension EnvironmentValues {
    var navigationManager: NavigationManager {
        get { self[NavigationManagerKey.self] }
        set { self[NavigationManagerKey.self] = newValue }
    }
}